### PR TITLE
[Merged by Bors] - feat: Add CategoryTheory.Sites.Coverage

### PR DESCRIPTION
**Pull request by adamtopaz**
This PR adds the notion of a coverage, and constructs the natural Galois insertion between coverages and Grothendieck topologies. We also show that a `Type _`-valued presheaf is a sheaf for a coverage if and only if it is a sheaf for the associated Grothendieck topology.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/Coverage.lean [hunk 1]
```diff
@@ -0,0 +1,365 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+
+import Mathlib.CategoryTheory.Sites.SheafOfTypes
+import Mathlib.Tactic
```
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Sites/Coverage.lean:8]:
```suggestion
```

File: Mathlib/CategoryTheory/Sites/Coverage.lean [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+Authors: Adam Topaz
+-/
+
+import Mathlib.CategoryTheory.Sites.SheafOfTypes
+
+/-!
+
+# Coverages
+
+A coverage `K` on a category `C` is a set of presieve associated to every object `X : C`,
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Sites/Coverage.lean:13]:
```suggestion
A coverage `K` on a category `C` is a set of presieves associated to every object `X : C`,
```

File: Mathlib/CategoryTheory/Sites/Coverage.lean [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+called "covering presieves".
+This collection must satisfy a certain "pullback compatibility" condition, saying that
+whenever `S` is a covering presieve on `X` and `f : Y ⟶ X` is a morphism, then there exists
+some covering sieve `T` on `Y` such that `T` factors through `S` along `f`.
+
+The main difference between a coverage and a Grothendieck pretopology is that we *do not*
+require `C` to have pullbacks.
+This is useful, for example, when we want to consider the Grothendieck topology on the category
+of extremally disconnected sets in the context of condensed mathematics.
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Sites/Coverage.lean:22]: What do you think of also mentioning the example of "(sub)basis for a topology"?

File: Mathlib/CategoryTheory/Sites/Coverage.lean [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+## Main Definitions and Results:
+
+All definitions are in th `CategoryTheory` namespace.
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Sites/Coverage.lean:26]:
```suggestion
All definitions are in the `CategoryTheory` namespace.
```

File: Mathlib/CategoryTheory/Sites/Coverage.lean [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+  · apply saturate.top
+  · assumption
+
+variable (C) in
+/--
+The Grothendieck topology associated to a coverage `K`.
+It is defined *inductively* as follows:
+1. If `S` is a covering presieve for `K`, then the sieve generated by `S` is a covering
+  sieve for the associated Grothendieck topology.
+2. The top sieves are in the assocaited Grothendieck topology.
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Sites/Coverage.lean:204]:
```suggestion
2. The top sieves are in the associated Grothendieck topology.
```

**Review by jcommelin — COMMENTED** [review]
Nice!

bors d+

File: Mathlib/CategoryTheory/Sites/Coverage.lean (hunk 2 above)
Comment (adamtopaz) [reply Mathlib/CategoryTheory/Sites/Coverage.lean]: Done!

File: Mathlib/CategoryTheory/Sites/Coverage.lean (hunk 4 above)
Comment (adamtopaz) [reply Mathlib/CategoryTheory/Sites/Coverage.lean]: Done

File: Mathlib/CategoryTheory/Sites/Coverage.lean (hunk 5 above)
Comment (adamtopaz) [reply Mathlib/CategoryTheory/Sites/Coverage.lean]: Done

File: Mathlib/CategoryTheory/Sites/Coverage.lean (hunk 3 above)
Comment (adamtopaz) [reply Mathlib/CategoryTheory/Sites/Coverage.lean:22]: Good idea! I'll add this later.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**